pkg/metricsadapter/controller.go

```diff
@@ -0,0 +1,166 @@
+package metricsadapter
+
+import (
+	"reflect"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	clusterV1alpha1 "github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1"
+	informerfactory "github.com/karmada-io/karmada/pkg/generated/informers/externalversions"
+	clusterlister "github.com/karmada-io/karmada/pkg/generated/listers/cluster/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/util"
+	"github.com/karmada-io/karmada/pkg/util/fedinformer/genericmanager"
+	"github.com/karmada-io/karmada/pkg/util/gclient"
+)
+
+// MetricsController is a controller for metrics, control the lifecycle of multi-clusters informer
+type MetricsController struct {
+	InformerFactory informerfactory.SharedInformerFactory
+	ClusterLister   clusterlister.ClusterLister
+	InformerManager genericmanager.MultiClusterInformerManager
+
+	queue      workqueue.RateLimitingInterface
+	restConfig *rest.Config
+}
+
+// NewMetricsController creates a new metrics controller
+func NewMetricsController(restConfig *rest.Config, factory informerfactory.SharedInformerFactory) *MetricsController {
+	clusterLister := factory.Cluster().V1alpha1().Clusters().Lister()
+	controller := &MetricsController{
+		InformerFactory: factory,
+		ClusterLister:   clusterLister,
+		InformerManager: genericmanager.GetInstance(),
```

Comment:
This MultiClusterInformerManager instance use `PodTransformFunc` to remove [some unnecessary fields](https://github.com/karmada-io/karmada/blob/master/pkg/util/fedinformer/transform.go#L72-L74),
However, HPA controller will use `PodCondition` to[ group pods](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/replica_calculator.go#L398-L402).

Reply:
For the metrics adapter, we only return the metrics and not the PodCondition. Therefore, it is acceptable. The HPA controller should be responsible for ensuring that everything is working correctly. Based on our testing, it appears to be functioning properly. @Poor12 will soon provide a PR for us to review.

Reply:
If metrics-controller only need to list pod and node, it's better to use typedInformer instead of dynamicInformer for better performance.

Reply:
We need metrics client, so we need genericmanager, not a typed manager, we can do optimization in the future.